### PR TITLE
GC: Implement eqref and i31ref

### PR DESCRIPTION
**Pull request by dcodeIO**

### CODE REVIEW

**Comment by dcodeIO** [comment]
So far this enables the `eqref` and `i31ref` types, and implements `ref.eq` so these can be tested.

Still a few TODOs for `i31ref`, and no `i31.new` etc. or meaningful tests just yet, because there is no story for a trivial constant `i31ref` literal (other nullable types can just use `ref.null`, but `i31ref` can't). Most likely requires refactoring literals so these can represent 1) a constant value `i31ref` and 2) an unknown value `i31ref` (i.e. imported). One solution might be to reuse the `.i32` field of literals and utilize `0x80000000` as a sentinel indicating an unknown value, but since other ref literals also need a story for known/unknown values, a more general solution might be preferrable.

File: src/wasm/wasm-validator.cpp [hunk 1]
```diff
@@ -1929,6 +1932,17 @@ void FunctionValidator::visitRefFunc(RefFunc* curr) {
   shouldBeTrue(!!func, curr, "function argument of ref.func must exist");
 }
 
+void FunctionValidator::visitRefEq(RefEq* curr) {
```
Comment (tlively) [inline src/wasm/wasm-validator.cpp:1935]: This should probably check that GC is enabled as well.

**Review by tlively — COMMENTED** [review]
It might be a good idea to split adding the new types and GC feature off from adding any new instructions, just to keep the patch size down.

**Comment by dcodeIO** [comment]
I see, hmm. My thinking was that, in order to implement `i31ref`, the internal representation of literals needs to be updated to be able to represent them (as outlined above), which would be a change we'd want to test somehow, so at a minimum we'd need `ref.eq` and `i31.new`, while all the other instructions can be postponed as "nice to have later"s.

I can of course try to split this if you prefer, and keep things stubbed out, but I have a feeling that at the end of the day this will result in more reviewing work overall due to the overhead of looking at the same places multiple times. In general I think that biting the bullet once or twice during the upcoming GC work will safe you a lot of time and me a lot of headaches.

**Comment by tlively** [comment]
Yeah, I think splitting would be good. Personally, the largest amount of overhead comes from looking through large patches multiple times over the course of a review. You're right that we need instructions to test the functionality of the new types, but an initial patch that just adds the new types to all the necessary switches and in the JS/C APIs without implementing or testing any functionality would be fine as well. Splitting also helps get non-controversial code in faster and helps keep discussion of the interesting/complex parts focused (for example, figuring out how to represent `let` and non-nullable references).

**Pull request closed without merging**